records/2023-06-27.scalability.increasing.send.rates.md

```diff
@@ -0,0 +1,276 @@
+| 8/3/2023 3:30 PM  | 8/3/2023 3:41 PM   | 11 minutes | 40000              | 40k Emails x 1 User as Bulk | 7       | r5.large      | 73          | minReplicas: 3  maxReplicas: 73concurrency 4New Node Size: r5.large  | resources:            requests:              cpu: "100m"              memory: "500Mi"            limits:              cpu: "550m"              memory: "1024Mi" | 5507           | 0.36 sec          |
+| 8/3/2023 3:50 PM  | 8/3/2023 4:06 PM   | 16 minutes | 80000              | 80k Emails x 1 User as Bulk | 7       | r5.large      | 73          | minReplicas: 3  maxReplicas: 73concurrency 4New Node Size: r5.large  | resources:            requests:              cpu: "100m"              memory: "500Mi"            limits:              cpu: "550m"              memory: "1024Mi" | 5603           | 0.36 sec          |
+
+## Decision: Scaling Version 1
+
+### Summary
+
+The initial scaling effort focussed on increasing the number of celery pods performing delivery-related work. We have settled on having the system scale up to 30 email delivery pods, 20 SMS delivery pods, and 30 core pods dedicated to all other celery tasks. The system also scales down to an idle state when appropriate. Other supporting work and changes were required to accomplish this while keeping the system stable and responsive.
+
+### Karpenter
+
+[Karpenter](https://aws.github.io/aws-eks-best-practices/karpenter/) is an open-source cluster autoscaler that automatically provisions new nodes in response to unschedulable pods. Essentially, when the cluster runs out of space karpenter will provision a new node for the cluster. Also, karpenter will occasionally decide that the running pods would fit on fewer nodes, and will remove a node from the cluster and redistribute the pods.
+
+The benefits of having karpenter scale our system up and down are:
+- we do not have to worry about running out of resources, and
+- we do not pay for computing power that we do not need.
+
+### Always available resources
+
+If we were to have no pods running at our idle state then there would be an unreasonable delay in sending new high priority messages. Thus we require a minimal number of pods to be available at all times. 
+
+In theory this should be possible with the kubernetes [horizontal pod autoscaler](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/) running on nodes managed by karpenter. In practice, however, we found that occasionally these scaling deployments would drop to zero pods for a few minutes. To prevent this from causing a delay our system has a group of nodes (called "on demand" nodes) that are always running. These nodes run our flask apps together with 3 each of the email, sms, and core celery pods.
+
+### Logging and metrics
+
+TODO: Logging and / or cloudwatch metrics got hosed. We added fluentbit to fix something?
```

Comment:
I think it came with an upgrade we had to do, not related per se to scaling up (maybe as a side effect of integrating karpenter? I don't remember but we could ask Ben. Nonetheless, we could get rid of a logging section, but we could mention that we have promoteus and other such tools on the metrics side maybe?